keps/sig-node/3815-pod-condition-for-configuration-errors/README.md

```diff
@@ -0,0 +1,1021 @@
+updates.
+
+[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
+-->
+
+As a batch user, it is very likely that users experience configuration errors when submitting pods to Kubernetes.  We would like to propose a new condition so that users can determine if a pod has an error in startup.  A new condition will provide a quick way to verify if a pod detected a unrecoverable configuration error.
+
+## Motivation
+
+Batch users may not have deep knowledge of Kubernetes.  It is common occurrence to find users submitting Pods with configuration errors (invalid images, wrong image names, missing volumes, missing secrets, wrong volume/secret for config map).  These cases cause the Pod to be stuck in pending and there is code required to remove these pods or user input to remove.
+
+As a batch system developer, we find that higher level controllers have to write code to handle Pending Pod cases.  Since the pods are not considered failed, the higher level controller needs to transition this state to failed in order to enforce retries.  For example, the job controller will run these pods and the state of the Pod will be in Pending.  The job controller does not transition this state to failed even if the pod will never start running.
+
+For CNCF projects, in the Armada project, we implement logic that determines retry ability based on events and statuses.  Events are not standard so we expect issues when updating and we discovered various states that don't have a representative status.  
+
+It would be ideal to have a standard condition for these common configuration errors.
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+### Goals
+- Add a new condition PodFailedToStartup to detect failures in configuration.
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+### Non-Goals
+- Will not modify behavior for ImagePullBackOff
+   - This status can be business as usual or image pull errors
+   - Unclear if it is possible to detect this as a failure
+   - Invalid ImagePullSecrets are a common fail case but it is difficult to 
+- Configuration errors that happen during ContainerCreation (Mounting volumes) will not be included as part of this condition
+   - PodReadyToStart condition actually covers these cases so there is no need to cover them in this KEP.
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+## Proposal
+
+- Add a PodFailedToStartup Condition for the following cases:
+
+  - Invalid Image Name
```

Comment:
What does this mean?  Is it runtime-specific, or should we be validating it at the API?

Reply:
If we validate this against a static specification, why don't we put that in API validation and just fail the pod / deployment / job / etc from being created?

@SergeyKanzhelev or @mrunalp do you know?

Reply:
Elsewhere we DO check for leading/trailing but only for pods and not pod templates (??).  As far as I can tell there is no other static validation on this field.

If kubelet is doing that in a runtime-specific way, then OK.  But if it is validating against a static rule, we should do it in regular API validation.

Reply:
This sounds like a potential useful addition. Should we maybe move that check into a separate KEP? 

I think it’s a useful feature but may be considered a breaking change so we should probably do validation separately from this KEP?

Reply:
https://github.com/kubernetes/kubernetes/pull/115736 for an example of how we put this validation in pod validation rather than in kubelet when pulling the image.

Reply:
I don't think it needs a KEP if we can assert that:

a) Today, there's no way to create a pod with an "invalid" image and NOT have it fail (IOW - any pod with an invalid name WILL fail to run)
b) This change just converts a 100% "runtime" hard-fail to an "api time" hard-fail.

Reply:
> a) Today, there's no way to create a pod with an "invalid" image and NOT have it fail

This is not accurate:

```
apiVersion: v1
kind: Pod
metadata:
  name: init-image-tagging-example-3
spec:
  volumes:
  - name: rootfs
    hostPath:
      path: /
  containers:
  - name: pause
    image: registry.k8s.io/pause:3.1
  initContainers:
  - name: init-myservice
    securityContext:
      privileged: true
    volumeMounts:
    - mountPath: /rootfs
      name: rootfs
      readOnly: true
    image: registry.k8s.io/e2e-test-images/agnhost:2.40
    command: ["nsenter", "--mount=/rootfs/proc/1/ns/mnt", "--", "sh", "-c"]
    args:
      - |
        ctr -n k8s.io images pull docker.io/library/hello-world:latest
        ctr -n k8s.io images tag --force docker.io/library/hello-world:latest localhost/image:tag
        read -t 5 -p "I am going to wait for 50 seconds only ..."
  - name: init-2
    image: localhost/image:tag
    imagePullPolicy: Never
```

Reply:
I believe that @smarterclayton has pointed out that this logic is not safe.  I will resolve as we now realize that we can't really do validation

Reply:
Going to resolve this as this idea was pointed out to be flawed.